### PR TITLE
Kill clients with error instead of sending unsupported protocol messages

### LINKED ISSUE #2343: Too easy to send unsupported events
If a client binds to a version of a protocol that is older than what we support, sending newer events will generally crash the client. We should not do this. As #2341 demonstrates it is quite easy to do this accidentally. We should consider changing the generated code to either make this harder, or make it harmless.

### PR DESCRIPTION
**Pull request by wmww**
Fixes #2343

### CODE REVIEW

File: src/wayland/wayland_base.cpp [hunk 1]
```diff
@@ -151,3 +151,23 @@ void mw::internal_error_processing_request(wl_client* client, char const* method
...
+    wl_client_post_implementation_error(
+        client,
+        "Mir tried to send %s@%d.%s to object with version %d (requires version %d)",
+        wl_resource_get_class(resource),
+        wl_resource_get_id(resource),
+        event,
+        wl_resource_get_version(resource),
+        required_version);
+
+    log_error(
```
Comment (RAOF) [inline src/wayland/wayland_base.cpp:166]: Nit: Maybe `log_critical`? Getting here means that we've hit an application-breaking Mir logic bug.

**Review by RAOF — APPROVED** [review]
LGTM.

Feel free to switch to `log_critical`, but I won't need to re-review for that.

bors delegate=wmww

**Comment by wmww** [comment]
bors r=RAOF

**Comment by Saviq** [comment]
bors retry

**Pull request merged**